crates/matrix-sdk-base/src/sliding_sync.rs

```diff
@@ -30,60 +30,73 @@ impl BaseClient {
+            extensions,
             // FIXME: missing compared to v3::Response
             //presence,
-            //account_data,
-            //to_device,
-            //device_lists,
-            //device_one_time_keys_count,
-            //device_unused_fallback_key_types,
             ..
         } = response;
 
-        // FIXME not yet supported by sliding sync. see
-        // https://github.com/matrix-org/matrix-rust-sdk/issues/1014
-        // #[cfg(feature = "encryption")]
-        // let to_device = {
-        //     if let Some(o) = self.olm_machine().await {
-        //         // Let the crypto machine handle the sync response, this
-        //         // decrypts to-device events, but leaves room events alone.
-        //         // This makes sure that we have the decryption keys for the room
-        //         // events at hand.
-        //         o.receive_sync_changes(
-        //             to_device,
-        //             &device_lists,
-        //             &device_one_time_keys_count,
-        //             device_unused_fallback_key_types.as_deref(),
-        //         )
-        //         .await?
-        //     } else {
-        //         to_device
-        //     }
-        // };
-
-        if rooms.is_empty() {
-            // nothing for us to handle here
+        if rooms.is_empty() && extensions.is_empty() {
+            // we received a room reshuffling event only, there won't be anything for us to
+            // process. stop early
             return Ok(SyncResponse::default());
         };
 
+        let v4::Extensions { to_device, e2ee, account_data, .. } = extensions;
+
+        let to_device_events = to_device.map(|v4| v4.events).unwrap_or_default();
+
+        #[cfg(feature = "e2e-encryption")]
+        let to_device_events = {
+            if let Some(e2ee) = &e2ee {
+                if let Some(o) = self.olm_machine() {
+                    // Let the crypto machine handle the sync response, this
+                    // decrypts to-device events, but leaves room events alone.
+                    // This makes sure that we have the decryption keys for the room
+                    // events at hand.
+                    o.receive_sync_changes(
+                        to_device_events,
+                        &e2ee.device_lists,
+                        &e2ee.device_one_time_keys_count,
+                        e2ee.device_unused_fallback_key_types.as_deref(),
+                    )
+                    .await?
+                } else {
+                    to_device_events
+                }
+            } else {
+                to_device_events
+            }
+        };
+
+        let (device_lists, device_one_time_keys_count) = if let Some(e2ee) = e2ee {
+            (
+                e2ee.device_lists,
+                e2ee.device_one_time_keys_count.into_iter().map(|(k, v)| (k, v.into())).collect(),
+            )
+        } else {
+            Default::default()
+        };
```

Comment:
I would suggest to write this as follows, but that's a matter of personal opinions :-p (not tested, but the spirit is here).

```suggestion
        let (device_lists, device_one_time_keys_count) = e2ee.map(|e2ee| (
            e2ee.device_lists,
            e2ee.device_one_time_keys_count.into_iter().map(|(k, v)| (k, v.into())).collect(),
        )).unwrap_or_default();
```